### PR TITLE
azure file volume: allow provisioner to create stoarge account secret in different namespace

### LINKED ISSUE #47452: Azure Storage Driver should allow customization of secret location
From @liggitt :

In the process of provisioning a PV, it creates a secret in the namespace of the PVC containing the account name and key (password) needed to mount the PV. This secret is visible to a namespace-limited user.

Steps:

1. Run a cluster with
the azure cloud provider
dynamic PV provisioning enabled
a storage class specified for the azure file dynamic provisioner
kube-controller-manager set to --use-service-account-credentials=false (the default) OR with the persistent-volume-binder service account given permission to create secrets
2. As a namespace-limited user, create a PVC requesting the azure file storageclass

Result:
Depending on your authorization policy, a secret with the password for the storage account is created in the same namespace as the PVC.

Relevant source:
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_file/azure_provision.go#L161-L169
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_storage.go#L273
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_file/azure_util.go#L78

Note:
If the cluster is running with RBAC authorization only, and the controller manager is running with --use-service-account-credentials, the persistent-volume-binder role does not permit creation of secrets, so the Provision() operation would fail, and the credentials would not be exposed.

We should update the PVC controller so that the namespace where the secret is created is customizable.

### PR DESCRIPTION
**Pull request by rootfs**
**What this PR does / why we need it**:
allow provisioner to create stoarge account secret in different namespace
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47452 

**Special notes for your reviewer**:
@liggitt @brendandburns 

**Release note**:

```release-note
allow provisioner to create stoarge account secret in different namespace
```

### CODE REVIEW

File: pkg/volume/azure_file/azure_file.go [hunk 1]
```diff
@@ -104,6 +105,13 @@ func (plugin *azureFilePlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod
 	if err != nil {
 		return nil, err
 	}
+	nameSpace := pod.Namespace
+	if spec.PersistentVolume != nil {
+		ns, err := getSecretNamespace(plugin.host.GetKubeClient(), spec.PersistentVolume)
```
Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: mounters aren't looking up storage class data today... information needed to mount should be in the PV

Comment (rootfs) [inline pkg/volume/azure_file/azure_file.go]: Existing mounters still get secrets from pod namespace

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: I'm referring to the information flow. This has the mounter resolving a storage class from a PV, which isn't done today.

Comment (rootfs) [inline pkg/volume/azure_file/azure_file.go]: Mounter has to work with new provisioner to pick up custom namespace. Existing mounter won't be able to look up secret in other namespace.

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: The provisioner has to propagate information the mounter needs into the PV

**Comment by liggitt** [comment]
additionally, the secret creation in SetAzureCredentials is still problematic. the provisioner is not given write permission to secrets in default cluster policy

**Comment by rootfs** [comment]
> additionally, the secret creation in SetAzureCredentials is still problematic. the provisioner is not given write permission to secrets in default cluster policy

@liggitt nice point! just added.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go [hunk 2]
```diff
@@ -191,6 +191,8 @@ func init() {
 			rbac.NewRule("get").Groups(legacyGroup).Resources("secrets").RuleOrDie(),
 			// openstack
 			rbac.NewRule("get", "list").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
+			// Azure File
+			rbac.NewRule("get", "create", "delete").Groups(legacyGroup).Resources("secrets").RuleOrDie(),
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go]: no, I wasn't saying the provisioner should gain this permission

Comment (rootfs) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go]: what's the right way?

File: pkg/volume/azure_file/azure_file.go (hunk 1 above)
Comment (rootfs) [inline pkg/volume/azure_file/azure_file.go]: @liggitt I haven't found that pattern of putting secret namespace in PV yet, the existing [PV annotations](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/volumehelper/volumehelper.go#L30) don't have one for secret namespace, while we have precedence of storage class name resolution in [mounter](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L159) and [deleter](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/storageos/storageos.go#L175)

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: Mounters run on the nodes, and have to be authorized to obtain secrets needed for mounting. This is done by the node authorizer based on https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolume/util.go#L33

To know which secrets a PV grants access to, we need structured data in the PV, not ad-hoc storage class parameters.

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: create/delete and attach/detach are done by controllers living outside the nodes, so it is more reasonable to grant broader permission to them, and/or have them resolve storage class info.

mount/unmount is done by arbitrary nodes, so segmenting secret access is far more important

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: also, the entire block you referenced will fail on any cluster running RBAC, as nodes do not have permission to create arbitrary endpoints/services

Comment (liggitt) [inline pkg/volume/azure_file/azure_file.go]: opened https://github.com/kubernetes/kubernetes/issues/47657 to track the issue with that glusterfs mount operation

**Comment by rootfs** [comment]
closing, reworking on a different way

**Pull request closed without merging**